### PR TITLE
add multi-modal (vision + language) transformers

### LINKED ISSUE #2775: Add multi-modality transformers
**Is your feature request related to a problem? Please describe.**
Add transformers that can be leveraged for processing multi-modal data (i.e. vision and language). The transformer block can be also used for creating cross-attention modules. 
**Describe the solution you'd like**
The architecture can be imported from HuggingFace.

### PR DESCRIPTION
**Pull request by ahatamiz**
Signed-off-by: ahatamizadeh <ahatamizadeh@nvidia.com>

Fixes #2775 #2776 #2777 .

### Description
This pull request adds the full pipeline and support for multimodal (vision + language ) transformers. The transformers implementation follow Huggingface  repository. 

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/networks/nets/vltransformer.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+import math
+import os
+import shutil
+import tarfile
+import tempfile
+from typing import Sequence, Union
+
+import torch
+import transformers
```
Comment (wyli) [inline monai/networks/nets/vltransformer.py:20]: these should be `optional_import`

File: monai/networks/nets/vltransformer.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+        num_mixed_layers: int = 2,
+    ) -> None:
+        """
+        Args:
+            num_language_layers: number of language transformer layers.
+            num_vision_layers: number of vision transformer layers.
+            num_mixed_layers: number of mixed transformer layers.
+        """
+        super().__init__()
+        bert_config = type(
```
Comment (wyli) [inline monai/networks/nets/vltransformer.py:225]: please make it a module-level variable `, like https://github.com/Project-MONAI/MONAI/blob/c50235c85dca411e878c0c3508eb27a0a76b37fb/monai/networks/nets/efficientnet.py#L28

File: setup.cfg [hunk 3]
```diff
@@ -44,6 +44,7 @@ all =
     openslide-python==1.1.2
     pandas
     einops
+    transformers==4.10.2
```
Comment (wyli) [inline setup.cfg:47]: I think we can try to use the latest version for now, as we have the nightly test to detect any compatibility issues

**Review by wyli — COMMENTED** [review]
thanks, please some initial comments inline

File: monai/networks/nets/vltransformer.py (hunk 1 above)
Comment (ahatamiz) [reply monai/networks/nets/vltransformer.py]: Sure, reflected

File: setup.cfg (hunk 3 above)
Comment (ahatamiz) [reply setup.cfg]: Sure, reflected

File: monai/networks/nets/vltransformer.py (hunk 2 above)
Comment (ahatamiz) [reply monai/networks/nets/vltransformer.py:229]: One issue here is that module-level variable does not work properly when trying to initialize model weights from bert checkpoint in huggingface library.

**Comment by ahatamiz** [comment]
@wyli it seems like some check fails due to the following which is not related to multi-modality ?

```
/home/runner/work/MONAI/MONAI/monai/data/grid_dataset.py:144:13: B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling.  See https://docs.python.org/3/tutorial/errors.html#exception-chaining for details.
/home/runner/work/MONAI/MONAI/monai/data/utils.py:286:9: B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling.  See https://docs.python.org/3/tutorial/errors.html#exception-chaining for details.
/home/runner/work/MONAI/MONAI/monai/data/utils.py:297:9: B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling.  See https://docs.python.org/3/tutorial/errors.html#exception-chaining for details.
/home/runner/work/MONAI/MONAI/monai/utils/aliases.py:74:13: B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling.  See https://docs.python.org/3/tutorial/errors.html#exception-chaining for details.
/home/runner/work/MONAI/MONAI/monai/transforms/inverse_batch_transform.py:102:13: B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling.  See https://docs.python.org/3/tutorial/errors.html#exception-chaining for details.
/home/runner/work/MONAI/MONAI/monai/_extensions/loader.py:37:9: B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling.  See https://docs.python.org/3/tutorial/errors.html#exception-chaining for details.
/home/runner/work/MONAI/MONAI/tests/utils.py:77:9: B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling.  See https://docs.python.org/3/tutorial/errors.html#exception-chaining for details.
7     B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling.  See https://docs.python.org/3/tutorial/errors.html#exception-chaining for details.
```

**Comment by wyli** [comment]
yes, I'm merging a fix for that flake8 issue very soon https://github.com/Project-MONAI/MONAI/pull/2940

**Comment by ahatamiz** [comment]
Hi @wyli 

Another issue that comes up is with the optional import ( or just import). For some reason `transformers` is not being installed and cannot be imported:

`monai.utils.module.OptionalImportError: from transformers.models.bert.modeling_bert import BertEmbeddings (No module named 'transformers').`

Thanks

**Comment by ahatamiz** [comment]
> yes, I'm merging a fix for that flake8 issue very soon #2940

Thanks much.

**Comment by wyli** [comment]
> BertEmbeddings

how to replicate this issue?

I tried and the following works
```python
Python 3.7.10 (default, Feb 26 2021, 10:16:00) 
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from monai.utils import optional_import
>>> bert_embeddings = optional_import("transformers.models.bert.modeling_bert", name="BertEmbeddings")[0]
>>> bert_embeddings
<class 'transformers.models.bert.modeling_bert.BertEmbeddings'>
>>> 
```

File: monai/networks/nets/vltransformer.py (hunk 2 above)
Comment (wyli) [reply monai/networks/nets/vltransformer.py:229]: how to replicate this issue, could you please elaborate?

**Comment by ahatamiz** [comment]
> > BertEmbeddings
> 
> how to replicate this issue?
> 
> I tried and the following works
> 
> ```python
> Python 3.7.10 (default, Feb 26 2021, 10:16:00) 
> [Clang 10.0.0 ] :: Anaconda, Inc. on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> >>> from monai.utils import optional_import
> >>> bert_embeddings = optional_import("transformers.models.bert.modeling_bert", name="BertEmbeddings")[0]
> >>> bert_embeddings
> <class 'transformers.models.bert.modeling_bert.BertEmbeddings'>
> >>> 
> ```

Thanks @wyli. I pushed a new PR based on this for optional imports.

**Comment by ahatamiz** [comment]
> > > BertEmbeddings
> > 
> > 
> > how to replicate this issue?
> > I tried and the following works
> > ```python
> > Python 3.7.10 (default, Feb 26 2021, 10:16:00) 
> > [Clang 10.0.0 ] :: Anaconda, Inc. on darwin
> > Type "help", "copyright", "credits" or "license" for more information.
> > >>> from monai.utils import optional_import
> > >>> bert_embeddings = optional_import("transformers.models.bert.modeling_bert", name="BertEmbeddings")[0]
> > >>> bert_embeddings
> > <class 'transformers.models.bert.modeling_bert.BertEmbeddings'>
> > >>> 
> > ```
> 
> Thanks @wyli. I pushed a new PR based on this for optional imports.

Hi @wyli , seems like this did not resolve the issue as some of the checks failed.

**Comment by wyli** [comment]
> > > > BertEmbeddings
> > > 
> > > 
> > > how to replicate this issue?
> > > I tried and the following works
> > > ```python
> > > Python 3.7.10 (default, Feb 26 2021, 10:16:00) 
> > > [Clang 10.0.0 ] :: Anaconda, Inc. on darwin
> > > Type "help", "copyright", "credits" or "license" for more information.
> > > >>> from monai.utils import optional_import
> > > >>> bert_embeddings = optional_import("transformers.models.bert.modeling_bert", name="BertEmbeddings")[0]
> > > >>> bert_embeddings
> > > <class 'transformers.models.bert.modeling_bert.BertEmbeddings'>
> > > >>> 
> > > ```
> > 
> > 
> > Thanks @wyli. I pushed a new PR based on this for optional imports.
> 
> Hi @wyli , seems like this did not resolve the issue as some of the checks failed.

Please add the test to https://github.com/Project-MONAI/MONAI/blob/76dc5c2774b46ee447ec11cb41ec834a0c84654c/tests/min_tests.py#L138

**Comment by ahatamiz** [comment]
> > > > > BertEmbeddings
> > > > 
> > > > 
> > > > how to replicate this issue?
> > > > I tried and the following works
> > > > ```python
> > > > Python 3.7.10 (default, Feb 26 2021, 10:16:00) 
> > > > [Clang 10.0.0 ] :: Anaconda, Inc. on darwin
> > > > Type "help", "copyright", "credits" or "license" for more information.
> > > > >>> from monai.utils import optional_import
> > > > >>> bert_embeddings = optional_import("transformers.models.bert.modeling_bert", name="BertEmbeddings")[0]
> > > > >>> bert_embeddings
> > > > <class 'transformers.models.bert.modeling_bert.BertEmbeddings'>
> > > > >>> 
> > > > ```
> > > 
> > > 
> > > Thanks @wyli. I pushed a new PR based on this for optional imports.
> > 
> > 
> > Hi @wyli , seems like this did not resolve the issue as some of the checks failed.
> 
> Please add the test to this list to exclude it for minimum environment tests
> 
> https://github.com/Project-MONAI/MONAI/blob/76dc5c2774b46ee447ec11cb41ec834a0c84654c/tests/min_tests.py#L138

Sure.

File: monai/networks/nets/vltransformer.py (hunk 2 above)
Comment (wyli) [reply monai/networks/nets/vltransformer.py:229]: or you can make the dictionary (in line 232-255) a global variable as the default, 
```py
def __init__(
        self,
        num_language_layers: int = 2,
        num_vision_layers: int = 2,
        num_mixed_layers: int = 2,
        bert_config: Dict = BERT_CONFIG,
    )
```
it'll leave some flexibility for tuning the parameters for the end users.

File: docs/requirements.txt [hunk 4]
```diff
@@ -20,3 +20,4 @@ sphinxcontrib-serializinghtml
 sphinx-autodoc-typehints==1.11.1
 pandas
 einops
+transformers==4.10.2
```
Comment (wyli) [inline docs/requirements.txt:23]:
```suggestion
transformers
```

File: tests/test_vltransformer.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                                        "num_language_layers": num_language_layers,
+                                        "num_classes": num_classes,
+                                        "drop_out": drop_out,
+                                    },
+                                    (2, num_classes),  # type: ignore
+                                ]
+                                TEST_CASE_VLTransformers.append(test_case)
+
+
+class TestPatchEmbeddingBlock(unittest.TestCase):
```
Comment (wyli) [inline tests/test_vltransformer.py:45]:
```suggestion
class TestVLTransformers(unittest.TestCase):
```

File: monai/networks/nets/vltransformer.py (hunk 2 above)
Comment (ahatamiz) [reply monai/networks/nets/vltransformer.py:229]: One way is to simply change the object-based definition to a dictionary-based config for BERT model, but `class BertPreTrainedModel ` is not compatible with this. I am using this class directly from Huggingface with very minor modifications. 

For your suggestion to give more flexibility, I think we need to keep the original BETR configurations to be able to use the pre-trained BERT weights (at least for this release). Pre-training a BERT type model on text data from scratch could be costly. Any suggestions ?

Comment (wyli) [reply monai/networks/nets/vltransformer.py:229]: are you sure that we need to keep exactly this version of config in order to load the pretrained?

Comment (ahatamiz) [reply monai/networks/nets/vltransformer.py:229]: Basically, I first tried with different configurations. But the model will not be initialized from the BERT pre-trained checkpoint,so takes longer for training.

Comment (wyli) [reply monai/networks/nets/vltransformer.py:229]: my concern is mainly about the hard-coded config. for example, as a user I want to set ``attention_probs_dropout_prob`: 0.2` and `"transformers_version": "4.10.1" for fine-tuning of the pretrained model, how can I set this `MultiModal` class without rewriting everything from scratch?

**Comment by ahatamiz** [comment]
I close this pull request due to many conflicts between the two branches. Will push a new one.

**Comment by wyli** [comment]
Hi @ahatamiz, why the close? the PR looks nice, sorry for the confusion that my previous comments might have created

**Comment by wyli** [comment]
the latest update with the flexible bert_config looks good to me.

**Comment by ahatamiz** [comment]
Hi @wyli .. dealing with so many conflicts in this PR was a bit time-consuming. I had to submit another one.
